server/src/main/java/io/crate/planner/optimizer/LoadedRules.java

```diff
@@ -32,71 +32,21 @@
-import io.crate.planner.optimizer.rule.MoveFilterBeneathGroupBy;
-import io.crate.planner.optimizer.rule.MoveFilterBeneathJoin;
-import io.crate.planner.optimizer.rule.MoveFilterBeneathOrder;
-import io.crate.planner.optimizer.rule.MoveFilterBeneathProjectSet;
-import io.crate.planner.optimizer.rule.MoveFilterBeneathRename;
-import io.crate.planner.optimizer.rule.MoveFilterBeneathUnion;
-import io.crate.planner.optimizer.rule.MoveFilterBeneathWindowAgg;
-import io.crate.planner.optimizer.rule.MoveOrderBeneathFetchOrEval;
-import io.crate.planner.optimizer.rule.MoveOrderBeneathNestedLoop;
-import io.crate.planner.optimizer.rule.MoveOrderBeneathRename;
-import io.crate.planner.optimizer.rule.MoveOrderBeneathUnion;
-import io.crate.planner.optimizer.rule.OptimizeCollectWhereClauseAccess;
-import io.crate.planner.optimizer.rule.RewriteNestedLoopJoinToHashJoin;
-import io.crate.planner.optimizer.rule.RemoveRedundantFetchOrEval;
-import io.crate.planner.optimizer.rule.RewriteFilterOnOuterJoinToInnerJoin;
-import io.crate.planner.optimizer.rule.RewriteGroupByKeysLimitToLimitDistinct;
-import io.crate.planner.optimizer.rule.RewriteToQueryThenFetch;
+import io.crate.planner.operators.LogicalPlanner;
 import io.crate.types.DataTypes;
 
 @Singleton
 public class LoadedRules implements SessionSettingProvider {
 
     private static final String OPTIMIZER_SETTING_PREFIX = "optimizer_";
 
-    private final List<Class<? extends Rule<?>>> rules = List.of(
-        RemoveRedundantFetchOrEval.class,
-        MergeAggregateAndCollectToCount.class,
-        MergeAggregateRenameAndCollectToCount.class,
-        MergeFilters.class,
-        MoveFilterBeneathRename.class,
-        MoveFilterBeneathFetchOrEval.class,
-        MoveFilterBeneathOrder.class,
-        MoveFilterBeneathProjectSet.class,
-        MoveFilterBeneathJoin.class,
-        MoveFilterBeneathUnion.class,
-        MoveFilterBeneathGroupBy.class,
-        MoveFilterBeneathWindowAgg.class,
-        MergeFilterAndCollect.class,
-        RewriteFilterOnOuterJoinToInnerJoin.class,
-        MoveOrderBeneathUnion.class,
-        MoveOrderBeneathNestedLoop.class,
-        MoveOrderBeneathFetchOrEval.class,
-        MoveOrderBeneathRename.class,
-        DeduplicateOrder.class,
-        OptimizeCollectWhereClauseAccess.class,
-        RewriteGroupByKeysLimitToLimitDistinct.class,
-        RewriteInsertFromSubQueryToInsertFromValues.class,
-        RewriteToQueryThenFetch.class,
-        RewriteNestedLoopJoinToHashJoin.class
-    );
-
     @Override
     public List<SessionSetting<?>> sessionSettings() {
-        return Lists2.map(rules, this::buildRuleSessionSetting);
+        return Lists2.map(LogicalPlanner.optimizerRules(), x -> buildRuleSessionSetting(x.getClass()));
     }
 
     @VisibleForTesting
-    SessionSetting<?> buildRuleSessionSetting(Class<? extends Rule<?>> rule) {
+    SessionSetting<?> buildRuleSessionSetting(Class<? extends Rule> rule) {
```

Comment:
`rule.getClass` yields `Class<? extends Rule>` instead of `Class<? extends Rule<?>>` from `Rule.class`. Makes no difference in Practice.

Reply:
Could be solved by:
```suggestion
    SessionSetting<?> buildRuleSessionSetting(Rule<?> rule) {
        var clazz = rule.getClass();
        var simpleName = clazz.getSimpleName();
        ....
```

Reply:
As mentioned above, `rule.getClass()` will return `Class<? extends Rule>` not  `Class<? extends Rule<?>>`, therefore this will not resolve it.

Reply:
I think it would still make sense to at least limit the warning to the place where it's an issue.

E.g. do something like:

```diff

     @Override
+    @SuppressWarnings("unchecked")
     public List<SessionSetting<?>> sessionSettings() {
-        return Lists2.map(LogicalPlanner.optimizerRules(), x -> buildRuleSessionSetting(x.getClass()));
+        return Lists2.map(LogicalPlanner.optimizerRules(), x -> buildRuleSessionSetting((Class<? extends Rule<?>>) x.getClass()));
     }
```

And keep using `Class<? extends Rule<?>> rule` everywhere else. Otherwise the warnings are scattered all over the place.


> What is the difference between Rule<?> and Rule ? Rule<?> expands to Rule<? extends Object> and Rule expands to Rule<Object>. The difference in the boundary of these types are minimal.

In this case it may not matter much, but there is a difference between `Rule<?>` and a raw `Rule`. In the latter the type erasue also maps to methods. An example:

```
    class Foo<T> {

        private List<String> names() {
            return List.of("x");
        }
    }

    @Test
    public void test_foo() throws Exception {
        Foo x = new Foo();

        // x.names() is _not_ List<String> because Foo is a raw type
        List names = x.names();
    }
```

Reply:
Thanks for the feedback, let's go with a cast then.

Reply:
Btw, I was indeed not aware that the raw type would also erase the types of the methods generics.